drivers: pwm: small improvements/fixes

gmarull opened the pull request:
This PR introduces some small improvements/fixes to PWM drivers

NOTE: removed all formatting for now until it is discussed, this PR contains now tiny fixes or enhancements.

gmarull committed:
drivers: pwm: gecko: fix non-local includes

HAL includes are not local, so use <> instead of "".

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pwm: imx: remove redundant DEV_BASE accessor

The base variable type stored in config already has the right type. Such
kind of accessors have been removed in many other places.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pwm: mchp_xec: fix include list

Include what is needed, and sort includes properly.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pwm: mcux: fix register casting

Register was wrongly casted to (void *) in one case, and unnecessarily
casted in another case.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pwm: add log level to LOG_MODULE_REGISTER

LOG_MODULE_REGISTER accepts log level as an argument.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pwm: npcx: use register type

Custom accessors like HAL_INSTANCE() have been gradually removed
in-tree. Instead, store a pointer with the right type (struct pwm_reg
*).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pwm: npcx: remove unused include <assert.h>

<assert.h> is not used, so remove it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: pwm: minor formatting enhancements

Some formatting tweaks to improve the outcome of clang-format.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

dkalowsk reviewed (dismissed):
Looks good.  I like the push towards clang-format usage.

gmarull committed:
scripts: checkpatch: add Atmel Pwm typedef

Add Atmel Pwm typedef so that when doing `Pwm * const pwm` definition
checkpatch doesn't complain about the position of the '*'. If not added,
we must define `Pwm *const pwm` to make checkpatch happy, but,
clang-format results in `Pwm * const pwm` (correct).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

henrikbrixandersen commented on drivers/pwm/pwm_b91.c:
```diff
@@ -6,18 +6,19 @@
 PINCTRL_DT_INST_DEFINE(0);
 
-static const struct pinctrl_dev_config *pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(0);
+static const struct pinctrl_dev_config *pcfg =
+	PINCTRL_DT_INST_DEV_CONFIG_GET(0);
```
Picking one change at random here: What is the gain of reformatting all the files unconditionally? The code does not get more readable by this.

Tools like `clang-format` can only work on a set of rules, but does not perceive code as the human brain does. Alignment of definitions on consecutive lines, if clauses, ... makes sense to the human brain and helps it see relationships, and aids in humans being able to scan the code quickly.

henrikbrixandersen replied:
FWIW, the Linux kernel process guidelines do not recommend blindly using `clang-format` to reformat either: https://www.kernel.org/doc/html/latest/process/clang-format.html

gmarull replied:
Auto-formatters are not perfect, this is why they can be turned off on certain regions if needed. Other languages have done a much better job at this from the beginning (e.g. Rust) and others like Python have great tools nowadays (e.g. black), C as usual is years behind. By looking at the Zephyr codebase, I still think an auto-formatter will help on areas like formatting device definition macros (it's now a mess). Auto-formatters are also useful when refactoring (it is painful without). The line above probably indicates that 80 cols is not the smartest choice for Zephyr. Maybe increasing to 100, or 88 (like did `black` in Python after a study) will yield better results.

gmarull replied:
> FWIW, the Linux kernel process guidelines do not recommend blindly using `clang-format` to reformat either: [kernel.org/doc/html/latest/process/clang-format.html](https://www.kernel.org/doc/html/latest/process/clang-format.html)

Linux Kernel coding style is loosely specified, that's probably why. `checkpatch` just checks a few things, but one can write code in subtle different styles while still making it pass... Linux is great in many aspects, but when it comes to coding consistency or documentation... there are better examples around, really.

henrikbrixandersen reviewed (changes requested):
NAK'ing until we have had further discussions on automatic code formatting.

gmarull commented on drivers/pwm/pwm_capture.c:
```diff
@@ -4,10 +4,9 @@
-#include <zephyr/zephyr.h>
 #include <zephyr/drivers/pwm.h>
 #include <zephyr/logging/log.h>
+#include <zephyr/zephyr.h>
```
it can go wrong if headers are not designed correctly (good rules can be found e.g. here https://google.github.io/styleguide/cppguide.html). In Zephyr there are a few cases where things need to be fixed, but from my experience most files will just work even if by chance.

nashif replied:
Looks like this is following some 'intelligent design' after all, so I might be wrong about this, but this is the stuff that might cause issues if not done correctly and left to a regex, so while interesting and nice and I agree in general with the cleanup, it should not be part of the linter and reformatting, maybe done as part of a targeted header cleanup instead.

dkalowsk replied:
While I like alphabetical headers, I don't think we want to enable that option just yet.  That is likely going to require a lot more scrutiny on the headers with tooling such as IWYU to clean them up.

nashif replied:
it is not alphabetical after all, it is following a more deliberare ordering, look at the .clang-config.

gmarull commented:
NOTE: removed all formatting for now until it is discussed, this PR contains now tiny fixes or enhancements.

erwango commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -720,8 +721,6 @@ replaced by 'st,prescaler' property in parent node, aka timers"
 									       \
 	static const struct pwm_stm32_config pwm_stm32_config_##index = {      \
 		.timer = (TIM_TypeDef *)DT_REG_ADDR(DT_INST_PARENT(index)),    \
-		/* For compatibility reason, use pwm st_prescaler property  */ \
-		/* if exist, otherwise use parent (timers) property         */ \
```
Is that required to clean this comment up?
If really required, can it be put elsewhere?

gmarull replied:
forgot to undo this one (clang-format gets confused with it). btw, shouldn't the `st,prescaler` at pwm level removed?

erwango replied:
Indeed, @ABOSTM can you check if it's now time to remove this?

ABOSTM replied:
Yes it is time to clean up,
we will take care of it.

gmarull commented:
@nashif ping, style changes are no longer part of the PR, please re-review
